streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatTransformValuesTest.java

```diff
@@ -68,69 +72,60 @@ public void shouldTransformInputRecordToMultipleOutputValues() {
                 "Hello",
                 "Blue",
                 "Planet");
+
         processor.init(context);
-        EasyMock.reset(valueTransformer);
 
-        EasyMock.expect(valueTransformer.transform(inputKey, inputValue)).andReturn(outputValues);
-        for (final String outputValue : outputValues) {
-            context.forward(new Record<>(inputKey, outputValue, 0L));
-        }
-        replayAll();
+        when(valueTransformer.transform(inputKey, inputValue)).thenReturn(outputValues);
 
         processor.process(new Record<>(inputKey, inputValue, 0L));
 
-        verifyAll();
+        for (final String outputValue : outputValues) {
+            verify(context).forward(new Record<>(inputKey, outputValue, 0L));
+        }
     }
 
     @Test
     public void shouldEmitNoRecordIfTransformReturnsEmptyList() {
         processor.init(context);
-        EasyMock.reset(valueTransformer);
 
-        EasyMock.expect(valueTransformer.transform(inputKey, inputValue)).andReturn(Collections.emptyList());
-        replayAll();
+        when(valueTransformer.transform(inputKey, inputValue)).thenReturn(Collections.emptyList());
 
         processor.process(new Record<>(inputKey, inputValue, 0L));
 
-        verifyAll();
+        verify(context, never()).forward(ArgumentMatchers.<Record<Integer, String>>any());
     }
 
     @Test
     public void shouldEmitNoRecordIfTransformReturnsNull() {
         processor.init(context);
-        EasyMock.reset(valueTransformer);
 
-        EasyMock.expect(valueTransformer.transform(inputKey, inputValue)).andReturn(null);
-        replayAll();
+        when(valueTransformer.transform(inputKey, inputValue)).thenReturn(null);
 
         processor.process(new Record<>(inputKey, inputValue, 0L));
 
-        verifyAll();
+        verify(context, never()).forward(ArgumentMatchers.<Record<Integer, String>>any());
     }
 
     @Test
     public void shouldCloseFlatTransformValuesProcessor() {
-        valueTransformer.close();
-        replayAll();
-
         processor.close();
 
-        verifyAll();
+        verify(valueTransformer).close();
     }
 
     @Test
     public void shouldGetFlatTransformValuesProcessor() {
+        @SuppressWarnings("unchecked")
         final ValueTransformerWithKeySupplier<Integer, Integer, Iterable<String>> valueTransformerSupplier =
             mock(ValueTransformerWithKeySupplier.class);
         final KStreamFlatTransformValues<Integer, Integer, String> processorSupplier =
             new KStreamFlatTransformValues<>(valueTransformerSupplier);
 
-        EasyMock.expect(valueTransformerSupplier.get()).andReturn(valueTransformer);
-        replayAll();
+        when(valueTransformerSupplier.get()).thenReturn(valueTransformer);
 
         final Processor<Integer, Integer, Integer, String> processor = processorSupplier.get();
 
-        verifyAll();
+        verify(valueTransformerSupplier).get();
```

Comment:
[optional]

I see that sometimes we explicitly verify stubbed actions and I'm wondering if there is a rule/guideline on when to do that (`stub` + `verify`) 🤔 

I'm asking because according to the [Mockito docs](https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#stubbing), if we stub something while using `Strict Stubbing mode`, we get verify part out-of-the-box.

<img width="597" alt="image" src="https://user-images.githubusercontent.com/9079844/184425984-5a23e722-83d0-4104-9428-dde40ba4fd67.png">

Reply:
I have a slight preference for not verifying something which has been explicitly stubbed. My rule of thumb is to only verify behaviour of mocks which would call important logic that we otherwise have no way of confirming purely by the return value(s). For example, verify that a `close()` method on some resource has been called. What is your opinion on this @cadonna? How would you like it approached?

Reply:
I did a bit more digging and I found https://github.com/mockito/mockito/issues/1086. As far as I understand the issue discussed there is that the documentation can be misinterpreted to say "if you use strict stubs verification happens without you using verifyNoMoreInteractions". This was fixed in https://github.com/mockito/mockito/pull/1280. However, it was un-fixed(?) again in https://github.com/mockito/mockito/commit/cf62c73422a14641ddcaad9c4a8f581daf899049. As far as I can tell the problem still persists in Mockito 4.7.0 (Kafka uses 4.4.0).

**This fails**:
```
    @After
    public void tearDown() {
        verifyNoMoreInteractions(transformer);
    }

    @Test
    public void shouldCallInitOfAdapteeTransformer() {
        when(transformerSupplier.get()).thenReturn(transformer);

        final TransformerSupplierAdapter<String, String, Integer, Integer> adapter =
            new TransformerSupplierAdapter<>(transformerSupplier);
        final Transformer<String, String, Iterable<KeyValue<Integer, Integer>>> adaptedTransformer = adapter.get();
        adaptedTransformer.init(context);

//        verify(transformerSupplier).get();
//        verify(transformer).init(context);
    }
```

**This succeeds**:
```
    @Test
    public void shouldCallInitOfAdapteeTransformer() {
        when(transformerSupplier.get()).thenReturn(transformer);

        final TransformerSupplierAdapter<String, String, Integer, Integer> adapter =
            new TransformerSupplierAdapter<>(transformerSupplier);
        final Transformer<String, String, Iterable<KeyValue<Integer, Integer>>> adaptedTransformer = adapter.get();
        adaptedTransformer.init(context);

//        verify(transformerSupplier).get();
//        verify(transformer).init(context);
    }
```

As such, I suspect that if we truly want to be as strict as possible we have to use both strict stubs and verifyNoMoreInteractions pretty much everywhere. Thoughts?

Reply:
@clolov I think we do not need to verify all stubs. In the case you pointed out above I think that verifying `verify(transformer).init(context)` is enough since the transformer is returned by the stub and if it were not called a would get a `NullpointerException`.

The following test would be fine with me:
```
@Test
public void shouldCallInitOfAdapteeTransformer() {
    when(transformerSupplier.get()).thenReturn(transformer);

    final TransformerSupplierAdapter<String, String, Integer, Integer> adapter =
        new TransformerSupplierAdapter<>(transformerSupplier);
    final Transformer<String, String, Iterable<KeyValue<Integer, Integer>>> adaptedTransformer = adapter.get();
    adaptedTransformer.init(context);

    verify(transformer).init(context);
}
```
However, there might be cases where it makes sense to verify the call of the stub.

I also think that we do not need to use `verifyNoMoreInteractions(transformer);` each and every time. We have a similar behavior with EasyMock and that leads to tests that are harder to maintain because even the slightest change in production code might lead to changes in the tests. We suffered a bit because of that in the past. Admittedly, without checking additional interactions you lose a bit of control in the test. I think it is a trade-off.   

What I do not understand is why the following test does not fail when run with `@RunWith(MockitoJUnitRunner.StrictStubs.class)`:

```
@Test
public void shouldCallInitOfAdapteeTransformer() {
    final Record record = mock(Record.class);
    when(transformerSupplier.get()).thenReturn(transformer);
    when(record.hasKey()).thenReturn(true);

    final TransformerSupplierAdapter<String, String, Integer, Integer> adapter =
        new TransformerSupplierAdapter<>(transformerSupplier);
    final Transformer<String, String, Iterable<KeyValue<Integer, Integer>>> adaptedTransformer = adapter.get();
    adaptedTransformer.init(context);

    verify(transformer).init(context);
}
``` 
`record.hasKey()` is never called and so the corresponding stubbing is unnecessary. According to Mockito's doc (https://javadoc.io/static/org.mockito/mockito-core/4.6.1/org/mockito/quality/Strictness.html#STRICT_STUBS):

>   Cleaner tests without unnecessary stubbings: the test fails when unused stubs are present (see UnnecessaryStubbingException).

Am I missing something?

Reply:
@cadonna I went into the source code of Mockito and came across the following
```
// only report when:
// 1. if all tests from given test have ran (filter requested is false)
//   Otherwise we would report unnecessary stubs even if the user runs just single test
// from the class
// 2. tests are successful (we don't want to add an extra failure on top of any existing
// failure, to avoid confusion)
```
Indeed, if you run the test class rather than just the test you do get the report. Why it does this, however, I do not know.

Reply:
I think it does so, because you could have stubbings in the setup method for the test class. If only some test methods use those stubbings, you would get failures if you run a single test that does not use the those stubbings.